Async method taking a reference argument fails to compile

I have this minimal example:
```rust
#[cfg(test)]
use mocktopus::macros::*;
pub struct Foo { }
#[cfg_attr(test, mockable)]
impl Foo {
    pub async fn read(&mut self, buf: &mut[u8]) {
        let _ = buf;
    }
}
```
It builds fine without tests, but fails with tests enabled with:
```
error[E0621]: explicit lifetime required in the type of `buf`
  --> src/main.rs:13:5
   |
12 |       pub async fn read(&mut self, buf: &mut[u8]) -> ()
   |                                         -------- help: add explicit lifetime `'mocktopus` to the type of `buf`: `&'mocktopus mut [u8]`
13 | /     {
14 | |         let _ = buf;
15 | |     }
   | |_____^ lifetime `'mocktopus` required
```
Using `pub async fn read<'a>(&mut self, buf: &'a mut[u8])` instead silences the error. I have no idea if it's a good workaround.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,8 @@
+## 0.7.7 - 2020-12-22
+- Transform async method with local binding
+- Fix missing lifetimes on async method refs
+## 0.7.6 - 2020-10-23
+- Transform all async methods
 ## 0.7.5 - 2019-11-22
 - Fix regression from 0.7.4
 - Fix potential name clash in generated code

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "mocktopus"
-version = "0.7.6"
+version = "0.7.7"
 authors = [
     "CodeSandwich <igor.zuk@protonmail.com>",
     "gregdhill <gregorydhill@outlook.com>"
@@ -21,7 +21,7 @@ travis-ci = { repository = "CodeSandwich/Mocktopus" }
 doctest = false
 
 [dependencies]
-mocktopus_macros = "=0.7.6"
+mocktopus_macros = "0.7.7"
 
 [dev-dependencies]
 tokio = { version = "0.2", features = ["full"] }

--- a/macros/Cargo.toml
+++ b/macros/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "mocktopus_macros"
-version = "0.7.6"
+version = "0.7.7"
 authors = [
     "CodeSandwich <igor.zuk@protonmail.com>",
     "gregdhill <gregorydhill@outlook.com>"

--- a/macros/src/item_injector.rs
+++ b/macros/src/item_injector.rs
@@ -259,7 +259,8 @@ fn inject_async_fn(
             predicates: Punctuated::new(),
         });
 
-    match outer_sig.inputs.iter_mut().next() {
+    let mut outer_sig_inputs = outer_sig.inputs.iter_mut();
+    match outer_sig_inputs.next() {
         Some(
             arg @ FnArg::Receiver(Receiver {
                 reference: Some(_), ..
@@ -293,6 +294,30 @@ fn inject_async_fn(
         }
         _ => {}
     };
+
+    for input in outer_sig_inputs {
+        if let arg @ FnArg::Typed(_) = input {
+            if let FnArg::Typed(PatType {
+                pat,
+                colon_token,
+                ty,
+                ..
+            }) = arg
+            {
+                if let Type::Reference(syn::TypeReference {
+                    and_token,
+                    lifetime,
+                    mutability,
+                    elem,
+                }) = *ty.clone()
+                {
+                    *arg = parse_quote! {
+                        #pat #colon_token #and_token 'mocktopus #lifetime #mutability #elem
+                    };
+                }
+            }
+        }
+    }
 
     outer_sig.generics.params.push(parse_quote!('life_self));
     for param in outer_sig.generics.params.iter() {
@@ -370,6 +395,11 @@ fn replace_self_in_stmt(stmt: &mut syn::Stmt) {
     match stmt {
         syn::Stmt::Semi(expr, _) => replace_self_in_expr(expr),
         syn::Stmt::Expr(expr) => replace_self_in_expr(expr),
+        syn::Stmt::Local(local) => {
+            if let Some((_, expr)) = &mut local.init {
+                replace_self_in_expr(expr);
+            }
+        }
         _ => (),
     }
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -260,6 +260,23 @@
 //! }
 //! ```
 //!
+//! ## Mocking async
+//! Mocking async functions is almost exactly the same as non-async:
+//!
+//! ```
+//! #[cfg_attr(test, mockable)]
+//! async fn sleep(ms: u64) {
+//!     tokio::time::delay_for(std::time::Duration::from_millis(ms)).await;
+//! }
+//!
+//! #[tokio::test]
+//! async fn sleep_test() {
+//!     sleep.mock_safe(|_| MockResult::Return(Box::pin(async move { () })));
+//!
+//!     sleep(10000).await;
+//! }
+//! ```
+//!
 //! # Mocking tricks
 //! ## Returning reference to value created inside mock
 //!

--- a/tests/mocking_methods_async/mod.rs
+++ b/tests/mocking_methods_async/mod.rs
@@ -1,5 +1,6 @@
 use super::*;
 
+mod when_struct_complex_method_regular_async;
 mod when_struct_generic_method_generic_async;
 mod when_struct_generic_method_regular_async;
 mod when_struct_regular_method_generic_async;

--- a/tests/mocking_methods_async/when_struct_complex_method_regular_async.rs
+++ b/tests/mocking_methods_async/when_struct_complex_method_regular_async.rs
@@ -1,0 +1,76 @@
+use super::*;
+
+struct Struct(u8);
+
+#[mockable]
+impl Struct {
+    async fn ref_method_with_binding(&self, arg: bool) -> String {
+        let method_call = || async { self.0 };
+
+        format!("{} {}", method_call().await, arg)
+    }
+
+    async fn ref_method_with_ref(&self, arg: &bool) -> String {
+        format!("{} {}", self.0, arg)
+    }
+}
+
+mod and_method_is_ref_method_with_binding {
+    use super::*;
+
+    #[tokio::test]
+    async fn and_not_mocked_then_runs_normally() {
+        assert_eq!("2 true", Struct(2).ref_method_with_binding(true).await);
+    }
+
+    #[tokio::test]
+    async fn and_continue_mocked_then_runs_with_modified_args() {
+        let struct_2 = Struct(2);
+        let struct_3 = Struct(3);
+        unsafe {
+            Struct::ref_method_with_binding.mock_raw(|_, b| MockResult::Continue((&struct_3, !b)));
+        }
+
+        assert_eq!("3 false", struct_2.ref_method_with_binding(true).await);
+        assert_eq!(2, struct_2.0);
+        assert_eq!(3, struct_3.0);
+    }
+
+    #[tokio::test]
+    async fn and_return_mocked_then_returns_mocking_result() {
+        let struct_2 = Struct(2);
+        unsafe {
+            Struct::ref_method_with_binding.mock_raw(|a, b| {
+                MockResult::Return(Box::pin(async move { format!("mocked {} {}", a.0, b) }))
+            });
+        }
+
+        assert_eq!(
+            "mocked 2 true",
+            struct_2.ref_method_with_binding(true).await
+        );
+        assert_eq!(2, struct_2.0);
+    }
+}
+
+mod and_method_is_ref_method_with_ref {
+    use super::*;
+
+    #[tokio::test]
+    async fn and_not_mocked_then_runs_normally() {
+        assert_eq!("2 true", Struct(2).ref_method_with_ref(&true).await);
+    }
+
+    #[tokio::test]
+    async fn and_return_mocked_then_returns_mocking_result() {
+        let struct_2 = Struct(2);
+        unsafe {
+            Struct::ref_method_with_ref.mock_raw(|a, b| {
+                MockResult::Return(Box::pin(async move { format!("mocked {} {}", a.0, b) }))
+            });
+        }
+
+        assert_eq!("mocked 2 true", struct_2.ref_method_with_ref(&true).await);
+        assert_eq!(2, struct_2.0);
+    }
+}

